jina/executors/indexers/keyvalue.py

```diff
@@ -75,6 +75,25 @@ def query(self, key: int) -> Optional[bytes]:
             with mmap.mmap(self.query_handler.body, offset=p, length=l) as m:
                 return m[r:]
 
+    def update(self, keys: Iterator[int], values: Iterator[bytes], *args, **kwargs):
+        for key, new_value in zip(keys, values):
+            pos_info = self.query_handler.header.get(key, None)
+            p, r, l = pos_info
+            with mmap.mmap(self.query_handler.body, offset=p, length=l) as m:
+                old_value = m[r:]
+                if old_value != new_value:
+                    # update
+                    l_end = len(new_value)  #: the length
+                    p_end = int(self._start / self._page_size) * self._page_size  #: offset of the page
+                    r_end = self._start % self._page_size  #
+                    self.write_handler.header.write(key, p_end, r_end, r_end + l_end)
+                    self.write_handler.body.write(new_value)
```

Comment:
as far as i understand, this tries to write the updated values in the *same* position but with varied length, if the length is longer than the old one, then the next entry in the `body` is overflowed.

Reply:
Ah yes we're not getting the l, p, r associated with the last entry.

Reply:
or is it simply appending, based on prev. `self._start`?

Reply:
but then why isn't `self._start` updated though?

anyway, a unit test on visiting all entries after a couple of updates would be clear enough to reveal the problem

Reply:
Can you check the new version?